### PR TITLE
drivers: ieee802154: add CSL receiver for nRF5

### PR DESCRIPTION
**Pull request by edmont**
Implement CSL receiver functionality in nRF5 radio driver.

Signed-off-by: Eduardo Montoya <eduardo.montoya@nordicsemi.no>

### CODE REVIEW

File: subsys/net/l2/openthread/Kconfig.thread [hunk 1]
```diff
@@ -88,7 +88,13 @@ config OPENTHREAD_CSL_SAMPLE_WINDOW
 
 config OPENTHREAD_CSL_RECEIVE_TIME_AHEAD
 	int "CSL receiver wake up margin in units of 10 symbols"
-	default 3
+	default 480
+
+config OPENTHREAD_CSL_MIN_RECEIVE_ON
+	int "Minimum CSL receive window"
+	default 5696
```
Comment (rlubos) [inline subsys/net/l2/openthread/Kconfig.thread:95]: Why such a weird value?

**Review by rlubos — APPROVED** [review]

File: subsys/net/l2/openthread/Kconfig.thread (hunk 1 above)
Comment (edmont) [reply subsys/net/l2/openthread/Kconfig.thread:95]: It comes from here: https://github.com/openthread/openthread/blob/main/src/core/config/mac.h#L409-L420

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 2]
```diff
@@ -766,6 +818,42 @@ static int nrf5_configure(const struct device *dev,
...
 
+	case IEEE802154_CONFIG_ENH_ACK_HEADER_IE: {
+		uint8_t short_addr_le[SHORT_ADDRESS_SIZE];
+		uint8_t ext_addr_le[EXTENDED_ADDRESS_SIZE];
+
+		/* We expect that addresses are in big endian format, but radio driver requires
+		 * to have them in little endian. Convert addresses to little endian.
+		 */
+		sys_put_le16(config->ack_ie.short_addr, short_addr_le);
+		sys_memcpy_swap(ext_addr_le, config->ack_ie.ext_addr, EXTENDED_ADDRESS_SIZE);
```
Comment (hubertmis) [inline drivers/ieee802154/ieee802154_nrf5.c:829]:
```suggestion
#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
		memcpy(ext_addr_le, config->ack_ie.ext_addr, EXTENDED_ADDRESS_SIZE);
#else
		sys_memcpy_swap(ext_addr_le, config->ack_ie.ext_addr, EXTENDED_ADDRESS_SIZE);
#endif
```

File: include/net/ieee802154_radio.h [hunk 3]
```diff
@@ -171,8 +172,8 @@ enum ieee802154_config_type {
 	/** Configure a radio reception slot */
```
Comment (hubertmis) [inline include/net/ieee802154_radio.h:172]: I think these options require more detailed comment describing that `RX_SLOT` can be used for any scheduler reception like Zigbee GP device, CSL, TSCH.

To implement CSL this option must be used together with the other two options. It should be explained when, why, and in what order these options must be set to maintain CSL receiver active.

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 4]
```diff
@@ -949,6 +1044,7 @@ static struct ieee802154_radio_api nrf5_radio_api = {
 	.tx = nrf5_tx,
 	.ed_scan = nrf5_energy_scan_start,
 	.get_time = nrf5_get_time,
+	.get_csl_acc = nrf5_get_acc,
```
Comment (hubertmis) [inline drivers/ieee802154/ieee802154_nrf5.c:1047]: I think this function could be named in a more generic way like `get_radio_clock_accuracy` or `get_radio_timings_accuracy`. It affects not only accuracy of CSL-related operations, but each scheduled TX or RX window timings.

**Review by hubertmis — CHANGES_REQUESTED** [review]
I think at least documentation how to use and implement CSL should be more descriptive before we merge it.

File: include/net/ieee802154_radio.h (hunk 3 above)
Comment (edmont) [reply include/net/ieee802154_radio.h:172]: Is the `rx_slot` struct as is compatible with all the mentioned protocols?

Comment (hubertmis) [reply include/net/ieee802154_radio.h:172]: I believe it is, but I have not tried to implement any. I might have overseen some missing details.

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 5]
```diff
@@ -766,6 +818,46 @@ static int nrf5_configure(const struct device *dev,
 		break;
 #endif /* CONFIG_NRF_802154_ENCRYPTION */
 
+	case IEEE802154_CONFIG_ENH_ACK_HEADER_IE: {
+		uint8_t short_addr_le[SHORT_ADDRESS_SIZE];
+		uint8_t ext_addr_le[EXTENDED_ADDRESS_SIZE];
+
+		/* We expect that addresses are in big endian format, but radio driver requires
+		 * to have them in little endian. Convert addresses to little endian.
+		 */
```
Comment (hubertmis) [inline drivers/ieee802154/ieee802154_nrf5.c:827]: Now this comment looks outdated.

File: include/net/ieee802154_radio.h [hunk 6]
```diff
@@ -168,11 +169,42 @@ enum ieee802154_config_type {
...
+	 *        Start CSL                                  Schedule CSL window
+	 *
+	 *    ENH_ACK_HEADER_IE                           CSL_RX_TIME (following window)
+	 *         |                                        |
+	 *         | CSL_RX_TIME (nearest window)           | RX_SLOT (nearest window)
+	 *         |    |                                   |   |
+	 *         |    | CSL_PERIOD                        |   |
+	 *         |    |    |                              |   |
+	 *         v    v    v                              v   v
+	 *    ----------------------------------------------------------[ CSL window ]-----
```
Comment (hubertmis) [inline include/net/ieee802154_radio.h:202]: It was just a sketch :laughing:

**Review by hubertmis — APPROVED** [review]

File: include/net/ieee802154_radio.h (hunk 6 above)
Comment (edmont) [reply include/net/ieee802154_radio.h:202]: But good enough to better understand the description :)

**Pull request merged**